### DISCUSSION TITLE
[Feature] Not able to find few apache skywalking metrics

### DISCUSSION THREAD

**Discussion by amolngt** [apache/skywalking#12784] [Q&A]
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/skywalking/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

In rabbitmq dashboard, in queue menu-> there is one column called "Message Rates"-> incoming, deliver / get and ack. I want to know metrics for each so that i can use that in rules for setting alarms. I found it for incoming, but its NOT working. Basically I am trying multiple metrics for rabbitmq.

### Use case

```
rules:
  # Alarm for Incoming Message Rate
  incoming_message_rate_high:
    metrics-name: meter_rabbitmq_messages_published
    threshold: 10
    period: 1
    count: 3
    silence-period: 5
    expression: meter_rabbitmq_messages_published > 10
    message: "High incoming message rate detected messages per second."

hooks:
  webhook:
    default:
      is-default: true
      urls:
        - http://localhost:3000/webhook # Replace with your webhook URL
```
I followed this link-> https://skywalking.apache.org/docs/main/next/en/setup/backend/backend-rabbitmq-monitoring/

But didnt get much help. Please let me know what metrics rule i can use for "Message Rates"-> incoming, deliver / get and ack. **Plus if you any other important metrics for rabbitmq, then please mention that also. **

Thank you in advance.

### Are you willing to submit a pull request to implement this on your own?

- [X] Yes I am willing to submit a pull request on my own!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by wu-sheng** [comment]
Change the dashboard into Edit mode. You could see the configurations of the widget, the metric name is there.

**Reply by amolngt** [reply]
Thank you sir for reply.

the configurations of the widget, which you are telling, there i can create a new widget. I want i metric names, which i can use for "Message Rates"-> incoming(rate at which incoming messages are coming), deliver / get(message consumed and fetched by application) and ack (slow ack) **for creating a rule** . And i want this for creating rules in alarm-settings.yml file.

see my complete alarm-settings.yml file.

```
rules:
  # Rule unique name, must be ended with `_rule`.
  endpoint_percent_rule:
    # A MQE expression and the root operation of the expression must be a Compare Operation.
    expression: sum((endpoint_sla / 100) < 75) >= 3
    # The length of time to evaluate the metrics
    period: 10
    # How many times of checks, the alarm keeps silence after alarm triggered, default as same as period.
    silence-period: 10
    message: Successful rate of endpoint {name} is lower than 75%

  service_percent_rule:
    expression: sum((service_sla / 100) < 85) >= 4
    period: 10
    silence-period: 10
    message: Service {name} successful rate is less than 85%

  service_resp_time_rule:
    expression: sum(service_resp_time >= 1000) >= 3
    # metrics-name: service_resp_time
    # op: ">"
    threshold: 1000 #ms
    period: 10
    # count: 3
    silence-period: 5
    message: "Response time of service {name} is over 1 second in 3 minutes."

  high_request_volume_rule:
    # metrics-name: service_cpm
    # threshold: 5
    # op: ">"
    period: 1
    # count: 3
    silence-period: 5
    expression: sum(service_cpm) >= 5 
    # only-as-condition: false
    # tags:
    #   level: WARNING
    message: "Service {name} is receiving a high volume of requests calls per minute."

  meter_rabbitmq_message_unacknowledged_delivered_consumers_rule:
    expression: sum(meter_rabbitmq_message_unacknowledged_delivered_consumers) > 100
    # op: ">"
    threshold: 100
    period: 10
    # count: 3
    silence-period: 5
    message: Service {name} has more than 100 Unacknowledged messages which are delivered_consumers in last 1 min  

  meter_rabbitmq_node_unacknowledged_messages_rule:
    expression: sum(meter_rabbitmq_node_unacknowledged_messages) > 100
    threshold: 100
    # op: ">"
    period: 1
    # count: 1
    silence-period: 5
    # only-as-condition: false
    # tags:
    #   level: WARNING
    message: "{name} has more than 10 unacknowledged messages."

  meter_rabbitmq_node_queue_messages_ready_rule:
    expression: sum(meter_rabbitmq_node_queue_messages_ready) > 100
    threshold: 100
    # op: ">"
    period: 1
    # count: 1
    silence-period: 2
    # only-as-condition: false
    # tags:
    #   level: WARNING
    message: "Ready messages available in Queue > 100"

  # Alarm for Incoming Message Rate
  **incoming_message_rate_high:**
    metrics-name: meter_rabbitmq_messages_published
    threshold: 10
    # op: ">"
    period: 1
    # count: 3
    silence-period: 5
    expression: meter_rabbitmq_messages_published > 10
    message: "High incoming message rate detected messages per second."

  # Alarm for Deliver/Get Message Rate
  **deliver_get_message_rate_high:**
    metrics-name: meter_rabbitmq_message_delivered_get
    threshold: 10
    # op: ">"
    period: 1
    # count: 3
    silence-period: 5
    expression: meter_rabbitmq_message_delivered_get > 10
    message: "High deliver/get message rate detected messages per second."

  # Alarm for Acknowledgment Message Rate
  **ack_message_rate_low:**
    metrics-name: meter_rabbitmq_message_ack
    threshold: 500
    # op: "<"
    period: 1
    # count: 3
    silence-period: 5
    expression: avg(meter_rabbitmq_message_ack) < 500
    message: "Low acknowledgment message rate detected messages per second."

hooks:
  webhook:
    default:
      is-default: true
      urls:
        - http://localhost:3000/webhook # Replace with your webhook URL 
```

from above code few rules are working and few not. **I want metrics for specifically for incoming_message_rate_high, deliver_get_message_rate_high & ack_message_rate_low. these 3 metrics i got from chatgpt**. which is not working. Let me know if you need more details.

**Reply by wu-sheng** [reply]
I don't actually know what do you mean. Why do you think a chatgpt can help you with that. 

There is no such magic and not complex.

**Reply by wu-sheng** [reply]
I would NOT be able to verify your YAML, as I don't have those metrics to set up. You should try to debug and find our which parts are not working.
We can only support you when you have some conclusions, rather than hope us to debug for you. That is not how we work.

**Reply by amolngt** [reply]
ok sir. so my basic question i am trying rules for alerting in apache-skywalking as per above code for "Message Rates"-> incoming(rate at which incoming messages are coming), deliver / get(message consumed and fetched by application) and ack (slow ack).
But it not working. can you give me the solution ?

**Reply by wu-sheng** [reply]
Do you see we build metrics for that? If not, it is an open question about nre metric. Not alerting.

**Reply by amolngt** [reply]
ok sir, can you tell me how can set rule queue wise for incoming messages / sec ?

**Reply by wu-sheng** [reply]
Like all metrics, you need to find that metrics on the UI widget. If you can't, you need to learn from OTEL and MAL step by step to build that new metric.

**Reply by amolngt** [reply]
hi sir,

Thank you for your reply.  I want to know that can rabbitmq-prometheus plugin gives metrics at queue level. so that i can use it in my apache skywalk alerts and will apache skywalk will support it ?

**Reply by wu-sheng** [reply]
That belongs to rabbit MQ project. We can't do anything unless the original source provided metrics.

You need to read their docs first, to see whether they provided this.

**Reply by amolngt** [reply]
![Screenshot from 2024-11-25 15-37-37](https://github.com/user-attachments/assets/65b45516-be74-4b6a-9fb9-b63baf039949)
![Screenshot from 2024-11-25 15-37-06](https://github.com/user-attachments/assets/50fe1096-9162-4fc3-9525-7fc9f0446bde)

if you above 2 screen shots rabbitmq-prometheus plugin gives us rabbitmq_queue_messages_unacked  and rabbitmq_channel_messages_delivered_total ... **queues wise**. But these metrics are not working in skywalk while creating rules. Is there anything we can do here ?

**Reply by wu-sheng** [reply]
You need to write MAL to process these metrics, then you have skywalking internal metrics.

- OTEL receiver, https://skywalking.apache.org/docs/main/latest/en/setup/backend/opentelemetry-receiver/
- MAL, https://skywalking.apache.org/docs/main/latest/en/concepts-and-designs/mal/

**Reply by amolngt** [reply]
hello sir,

i am getting metrics in rabbitmq-prometheus. can you give me 1 example(or any google link) for MAL to process  metrics? i am stuck. I am able to get default metrics. but metrics coming from rabbitmq-prometheus (http://localhost:15692/metrics/), i am not able to capture.

**Reply by wu-sheng** [reply]
I have provided all the documents. Many people followed that, and work well. Be patient. I feel you are too hurry and always want others to put things together for you.

**Reply by amolngt** [reply]
Hi sir,

I tried a lot. nowhere in you document has written about handling rabbitmq-prometheus metrics in OAP. 
there is a metrics called **rabbitmq_queue_messages_unacked** coming from rabbitmq-prometheus (http://localhost:15692/metrics/). **But otel-collector is not able to scrape it**. I am getting below error in OAP docker container->

java.lang.IllegalArgumentException: org.apache.skywalking.mqe.rt.exception.IllegalExpressionException: Expression: sum(rabbitmq_queue_messages_unacked) > 100 error: Metric: [rabbitmq_queue_messages_unacked] does not exist.

It would be really appreciating if you provide me some examples or as you said Many people followed might have got solution from your doc.

Thank you in advance.

**Reply by wu-sheng** [reply]
> But otel-collector is not able to scrape it. I am getting below error in OAP docker container

If so, don't you need to check otel collector? That is not SkyWalking project managed.

**Reply by amolngt** [reply]
Hi sir,
I am still working on same issue. Now i understood few things. rabbitmq-prometheus plugin is able to generate metrics like **rabbitmq_queue_messages_unacked** by addding below config in rabbitmq.conf file
```
prometheus.tcp.port = 15692
prometheus.return_per_object_metrics = true
```

but prometheus which is getting used inside skywalk- OAP (running on port 9090) is not able to scrape it. if i use separate prometheus on different port then it work, but OAP internal  prometheus doesn't allow that.
basically i want **rabbitmq_queue_messages_unacked** (queue wise messages count) metric which i can use in OAP alarm-settings to create a rule and send an alarm. but i get error(from OAP) ->

`java.lang.IllegalArgumentException: org.apache.skywalking.mqe.rt.exception.IllegalExpressionException: Expression: rabbitmq_queue_messages_unacked > 100 error: Metric: [rabbitmq_queue_messages_unacked] does not exist.`

Thank you in advance.

**Reply by wu-sheng** [reply]
If you add this into rabbitmq side, and want to forward all metrics, you should change the OTEL collector configuration.

**Reply by amolngt** [reply]
otel-collector gets installed after oap, because it wants oap connection. And inside oap volume, we give path of alarm-settings(rules to trigger alarm) file. So ideally oap should run first.
otel-collector-config.yaml file

```
receivers:
  prometheus:
    config:
      scrape_configs:
        - job_name: 'rabbitmq'
          metrics_path: "/metrics"
          scrape_interval: 5s
          static_configs:
            - targets: ['rabbitmq_ip:15692']

processors:
  batch: {}

exporters:
  otlp:
    endpoint: oap_ip:11800
    tls:
      insecure: true

service:
  pipelines:
    metrics:
      receivers: [prometheus]
      processors: [batch]
      exporters: [otlp]
```

inside docker-compose file for oap i have mentioned below volume
```
    volumes:
      - ./alarm-settings.yml:/skywalking/config/alarm-settings.yml
```

let me know if you need more details.

**Reply by wu-sheng** [reply]
Simply, you could read OTLP docs, you could show you whether this metric is collected and forwarded. such as adding a console exporter.

**Reply by wu-sheng** [reply]
Or you could open debug log of OAP. You would see received logs in the OAP terminal.

**Reply by amolngt** [reply]
Hi sir,
i read all docs, but not getting perfect info for metric like **rabbitmq_queue_messages_unacked**
can you tell me is it possible to scrape data from rabbitmq-prometheus for metric like **rabbitmq_queue_messages_unacked**. and can i create alarm-settings for it. if yes, can you share any github repo or pseudocode.

**Reply by wu-sheng** [reply]
I have nothing to do. It is what OTEL does and forwards. No magic.

**Reply by amolngt** [reply]
That's why i am saying .. can you share some doc or github link where people solved this issue. I have already read your doc, but didn't get any help.
Thank you in advance.

**Reply by wu-sheng** [reply]
Because that is not SkyWalking scope. That is OpenTelemetry project scope.

**Reply by amolngt** [reply]
Now i have my custom telemetry-collector, which is exporting metrics on xyz port. I want to know how to use **prometheus fetcher** and where to keep this config??

Thank you in advance.

**Reply by wu-sheng** [reply]
You could use otel collector to forward through otlp format.
We don't have Prometheus fetcher.

**Reply by amolngt** [reply]
Hi sir,
Have a look at my **otel-collector-config.yaml** file

```
receivers:
  prometheus:
    config:
      scrape_configs:
        - job_name: 'rabbitmq-monitoring'
          scrape_interval: 30s
          metrics_path: '/metrics'
          static_configs:
            - targets:
                - local_ip:15692
                - local_ip:8081
                # - rmq1:15692
                # - rmq2:15692
          relabel_configs:
            - source_labels: [ ]
              target_label: cluster
              replacement: rabbitmq-cluster
            - source_labels: [ __address__ ]
              regex: (.+)
              target_label: node
              replacement: $$1

processors:
  batch:

exporters:
  ####################################################################################
  # If you want to use otlp exporter please ensure that your OAP version is >= 9.2.0 #
  ####################################################################################
  otlp:
    endpoint: local_ip:11800
    tls:
      insecure: true
  logging:
    loglevel: debug

service:
  pipelines:
    metrics:
      receivers:
        - prometheus
      processors:
        - batch
      exporters:
        - otlp
        - logging
```

I am able to get metrics in  custom-exporter which is - local_ip:8081. But not able get same metrics in skywalk ui.
Please check attachments.
![Screenshot from 2024-12-17 18-57-30](https://github.com/user-attachments/assets/9289d090-d3d7-4757-873d-a2102ad0291e)
![Screenshot from 2024-12-17 18-56-50](https://github.com/user-attachments/assets/4cd6d265-7e19-46fc-86b5-b2ef84ab574c)


Is there any way , where i can check my metrics are scraped or loaded. let me know if you need more details.
thank you in advance.

**Reply by wu-sheng** [reply]
You should follow the doc, and learn MAL. 

- OTLP receiver doc, https://skywalking.apache.org/docs/main/latest/en/setup/backend/opentelemetry-receiver/
- MAL doc, https://skywalking.apache.org/docs/main/latest/en/concepts-and-designs/mal/

There are a lot of examples listed there. And that is how SkyWalking bundled dashboard works.

**Reply by amolngt** [reply]
docs are not providing relevant information. Nowhere it is written how to use custom exporter and metric.

**Reply by amolngt** [reply]
how can i edit oap's config which internally uses prometheus ? prometheus which is running on port 9090 its not taking rabbitmq metrics.

**Reply by wu-sheng** [reply]
Did you ever read what I wrote? I felt you just put words here, but ignore what I replied.
If so, what is the point.
Very frustrated to continue this thread.

**Reply by amolngt** [reply]
i am frustrated because i am not getting help from your docs since last 1 month. Please if you think my above question solutions is in your doc, then highlight that part and send me screenshot.

Thank you in advance.

**Reply by wu-sheng** [reply]
I said, you need otel to collect and forward metrics through otlp, and write MAL in oap.

You keep asking for something we don't have. In another one, I told you the mongo client version, then you asked again.

You can't stay in your own guessing about how this works.

**Reply by amolngt** [reply]
i did the same. But still it is not working. Please if you think my above question solutions is in your doc, then highlight that part and send me screenshot.

**Reply by wu-sheng** [reply]
Your question is exactly showing, you didn't read my reply seriously.

> how can i edit oap's config which internally uses prometheus ?

oap has nothing related to Prometheus.

That is my whole point.

**Reply by wu-sheng** [reply]
Otlp receiver is the one oap provided to process metrics, and that is the doc I share with otlp receiver.

But you keep asking for something doesn't exist.

**Reply by wu-sheng** [reply]
I totally feel like I am wasting my time replying. I repeated again, **you need otel to collect and forward metrics through otlp, and write MAL in oap.**
I am not going to reply anything if you keep ignoring my saying, but asking in your own imagination.

**Reply by amolngt** [reply]
at this location - /skywalking/config/otel-rules/rabbitmq/rabbitmq-cluster.yaml
i have edited rabbitmq-cluster.yaml file and added below code

```
  - name: queue_messages_unacked
    exp: rabbitmq_queue_messages_unacked
```

now i am able to get this metric. **but not able to get data.** 
can you tell what else i have to do?

**Reply by wu-sheng** [reply]
I think I also replied that, You need to ship this from otlp collector. Do you finish that part?

**Reply by amolngt** [reply]
i tried multiple expression, but all are not working. i want to get count and data queues wise for unacknowledged messages.
below  expressions are not working
```
exp: rabbitmq_queue_messages_unacked.filter(['queue']).sum(['cluster', 'queue'])
exp: rabbitmq_queue_messages_unacked{queue=~".*"}
```

Please help. **In documentation nowhere mentioned anything at queue level.**
can you send me any github discussion link where people have already done this?
Thank you in advance.
